src/Features/Core/Portable/Completion/CompletionItem.cs

```diff
@@ -89,7 +92,47 @@ public sealed class CompletionItem : IComparable<CompletionItem>
         /// <summary>
         /// Additional information attached to a completion item by it creator.
         /// </summary>
-        public ImmutableDictionary<string, string> Properties { get; }
+        public ImmutableDictionary<string, string> Properties
+        {
+            get
+            {
+                if (_propertiesAsImmutableDictionary is null)
+                    _propertiesAsImmutableDictionary = _properties.ToImmutableDictionary();
+
+                return _propertiesAsImmutableDictionary;
+            }
+        }
+
+        internal IEnumerable<KeyValuePair<string, string>> GetProperties()
+        {
+            return _properties;
+        }
+
+        internal bool TryGetProperty(string name, [NotNullWhen(true)] out string? value)
+        {
+            if (_propertiesAsImmutableDictionary is not null)
+                return _propertiesAsImmutableDictionary.TryGetValue(name, out value);
+
+            foreach ((var propName, var propValue) in _properties)
+            {
+                if (name == propName)
+                {
+                    value = propValue;
+                    return true;
+                }
+            }
+
+            value = null;
+            return false;
+        }
+
+        internal string GetProperty(string name)
+        {
+            if (TryGetProperty(name, out var value))
+                return value;
+
+            throw new KeyNotFoundException($"Property {name} not found");
```

Comment:
```suggestion
            // Allow ImmutableDictionary to throw the appropriate exception
            return Properties[name];
```

Reply:
I'm not sure how much value there is to getting the ImmutableDictionary to throw over our own throw. I definitely don't want to create the immutable dictionary to do so, and I think it would be weird to have different exception behavior for when the dictionary is created and when it isn't. I'm inclined not to change this unless you feel strongly.

Reply:
i'm fine with us throwing our own exception.